include/pybind11/pybind11.h

```diff
@@ -1116,24 +1116,38 @@ class class_ : public detail::generic_type {
 template <typename Type> class enum_ : public class_<Type> {
 public:
     using class_<Type>::def;
+    using class_<Type>::def_property_readonly_static;
     using Scalar = typename std::underlying_type<Type>::type;
     template <typename T> using arithmetic_tag = std::is_same<T, arithmetic>;
 
     template <typename... Extra>
     enum_(const handle &scope, const char *name, const Extra&... extra)
-      : class_<Type>(scope, name, extra...), m_parent(scope) {
+      : class_<Type>(scope, name, extra...), m_entries(), m_parent(scope) {
 
         constexpr bool is_arithmetic =
             !std::is_same<detail::first_of_t<arithmetic_tag, void, Extra...>,
                           void>::value;
 
-        auto entries = new std::unordered_map<Scalar, const char *>();
-        def("__repr__", [name, entries](Type value) -> std::string {
-            auto it = entries->find((Scalar) value);
-            return std::string(name) + "." +
-                ((it == entries->end()) ? std::string("???")
-                                        : std::string(it->second));
+        auto m_entries_ptr = m_entries.inc_ref().ptr();
+        def("__repr__", [name, m_entries_ptr](Type value) -> std::string {
+#if !defined(_MSC_VER)
+#define PREFIX_MEMBER_TEMPLATE template
+#else
+#define PREFIX_MEMBER_TEMPLATE
+#endif
+            for (const auto &kv : reinterpret_borrow<dict>(m_entries_ptr)) {
+                if (kv.second.PREFIX_MEMBER_TEMPLATE cast<Type>() == value)
+                    return std::string(name) + "." + kv.first.PREFIX_MEMBER_TEMPLATE cast<std::string>();
+            }
+            return std::string(name) + ".???";
+#undef PREFIX_MEMBER_TEMPLATE
```

Comment:
I think you can simply this whole block (and get rid of the ugly MSVC workaround macro) as:

```C++
        def("__repr__", [name, m_entries_ptr](Type value) -> str {
            for (const auto &kv : reinterpret_borrow<dict>(m_entries_ptr)) {
                if (pybind11::cast<Type>(kv.second) == value)
                    return str("{}.{}").format(name, kv.first);
            }
            return str("{}.???").format(name);
        });
```